tests/interop_client/main.hs

```diff
@@ -506,6 +516,83 @@ runClientStreamingTest opts = do
+      case decodeMessage (L.toStrict resp) of
+        Left err -> Left ("proto decoder says: " ++ err)
+        Right msg -> case _StreamingOutputCallResponse'payload msg of
+                       Nothing -> Left "no payload"
+                       Just payload -> if _Payload'body payload == expectedBody
+                                          then Right ()
+                                          else Left "payload does not match"
+      where
+        expectedBody = B.replicate (fromIntegral expectedSize) 0
+
+    assertResponses resps
+      | length resps == length responseSizes =
+        case errors of
+          [] -> Right ()
+          (err : _) -> Left err
+      | otherwise = Left ("number of received messages mismatch: " ++ show (length resps))
+      where
+        assertions = map assertResponse (responseSizes `zip` resps)
+        errors = lefts assertions
+  bracket (newChannel opts) destroyChannel $ \channel ->
+    bracket (newClientContext channel) destroyClientContext $ \ctx -> do
+      client <- callDownstream ctx callOptions "/grpc.testing.TestService/StreamingOutputCall" (encodeMessage req)
+      resps <- withNewClient client $ do
+        msgs <- receiveAllMessages
+        closeCall
+        return msgs
+      case resps of
+        RpcOk resps' ->
+          return (assertResponses resps')
+        RpcError err ->
+          return (Left (show err))
+
+runServerStreamingWithSlowConsumerTest :: Options -> IO (Either String ())
+runServerStreamingWithSlowConsumerTest opts = do
+  let
+    responseSize = 1030
+    responsesCount = 2000
+    delaySeconds = 20
+    delay = delaySeconds * 1000
+    req = def { _StreamingOutputCallRequest'responseParameters = map (\_ -> def { _ResponseParameters'size = responseSize }) [1 .. responsesCount] }
+    expectedBody = B.replicate (fromIntegral responseSize) 0
+    go acc = do
+      maybeResponse <- receiveMessage
+      case maybeResponse of
+        Nothing -> return $ Right (reverse acc)
+        Just resp ->
+          case decodeMessage (L.toStrict resp) of
+            Left err -> return $ Left ("proto decoder says: " ++ err)
+            Right msg -> case _StreamingOutputCallResponse'payload msg of
+                          Nothing -> return $ Left "no payload"
+                          Just payload -> if _Payload'body payload == expectedBody
+                                              then do liftIO $ threadDelay delay
+                                                      go (msg : acc)
+                                              else return $ Left "payload does not match"
+  bracket (newChannel opts) destroyChannel $ \channel ->
+    bracket (newClientContext channel) destroyClientContext $ \ctx -> do
+      client <- callDownstream ctx mempty "/grpc.testing.TestService/StreamingOutputCall" (encodeMessage req)
+      resp <- withNewClient client $ do
+        msgs <- go []
+        closeCall
+        return msgs
+      case resp of
+        RpcOk result ->
```

Comment:
This would be better written as;

```haskell
        RpcOk (Right msgs)
          | length msgs == responsesCount -> return (Right ())
          | otherwise -> return (Left ("responses count does not match: " ++ (show (length msgs))))
        RpcOk (Left err) ->
          return (Left err)
```